splicemachine/spliceengine
# DB-10436 Rollback to Transaction

New PR after the revert of https://github.com/splicemachine/spliceengine/pull/5103

Jiras following up on discussion in the original PR:

Failure Handling
https://splicemachine.atlassian.net/browse/DB-11394

Handling Running Transactions
https://splicemachine.atlassian.net/browse/DB-11332

Including Flush in Major Compaction
https://splicemachine.atlassian.net/browse/DB-11395

Also, a dependency discovered today that it's not possible to restart Olap Server from java code in K8s:
https://splicemachine.atlassian.net/browse/DBAAS-5061

# Short Description
Allow rollback of entire DB to a past transaction ID.

# Long Description
The original motivation, as noted in the Jira, is that 
> Customer wants to run a whole bunch of script tests - each from the same known starting point over and over again.  So it stands to reason they would restore schema/database to that starting point and rerun

Recommended procedure when running rollback:
1. Connect to DB at snapshot of the transaction ID you want to rollback to.
Example: `connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd;snapshot=7936';`
2. Do sanity check of the DB at this snapshot to be sure it's in the desired state you want to rollback to.
3. Rollback to the transaction ID: `CALL SYSCS_UTIL.SYSCS_ROLLBACK_DATABASE_TO_TRANSACTION();`
4. Connect to current DB.
Example: `connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd';`
5. If running in a K8s environment, restart Olap Server.

Recommend copying and pasting the transaction ID to ensure accuracy.

Note: Requires enterprise access.

# How to test
Make several DB changes.  Get a transaction ID from among the changes.  Rollback to the transaction ID and verify the subsequent changes were rolled back.
Example:

```
create table a ( i int);
insert into a values 1,2;
select * from a;
CALL SYSCS_UTIL.SYSCS_GET_CURRENT_TRANSACTION();  -- 7936
select * from a;
create table b ( i int);
insert into a values 3,4;
select * from a;
select * from b;
CALL SYSCS_UTIL.SYSCS_ENABLE_ENTERPRISE('XXXX');

connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd;snapshot=7936';
-- Verify the DB is in the state you want
CALL SYSCS_UTIL.SYSCS_ROLLBACK_DATABASE_TO_TRANSACTION();
connect 'jdbc:splice://localhost:1527/splicedb;user=userid;password=abcd';

select * from a;
select * from b;
```

At the end, `select * from a;` should return 1 and 2.
`select * from b;` should return `ERROR 42X05: Table/View 'B' does not exist.`

More test examples are in the comments of the QA Jira: https://splicemachine.atlassian.net/browse/QA-2773
